core/src/sumeragi/view_change.rs

```diff
@@ -41,90 +42,92 @@ pub struct ProofBuilder(SignedProof);
 
 impl ProofBuilder {
     /// Constructor from index.
-    pub fn new(latest_block_hash: Option<HashOf<SignedBlock>>, view_change_index: u64) -> Self {
+    pub fn new(latest_block_hash: Option<HashOf<SignedBlock>>, view_change_index: usize) -> Self {
+        let view_change_index = view_change_index as u64;
+
         let proof = SignedProof {
             payload: ProofPayload {
                 latest_block_hash,
                 view_change_index,
             },
-            signatures: [].into_iter().collect(),
+            signatures: Vec::new(),
         };
 
         Self(proof)
     }
 
     /// Sign this message with the peer's public and private key.
-    pub fn sign(mut self, key_pair: &KeyPair) -> SignedProof {
-        let signature = SignatureOf::new(key_pair, &self.0.payload);
-        self.0.signatures.insert(signature);
+    pub fn sign(mut self, signatory_idx: usize, private_key: &PrivateKey) -> SignedProof {
+        let signature = SignatureOf::new(private_key, &self.0.payload);
+        self.0.signatures.push((signatory_idx as u64, signature));
         self.0
     }
 }
 
 impl SignedProof {
     /// Verify the signatures of `other` and add them to this proof.
-    fn merge_signatures(&mut self, other: SignaturesOf<ProofPayload>) {
-        for signature in other {
-            if signature.verify(&self.payload).is_ok() {
-                self.signatures.insert(signature);
+    fn merge_signatures(&mut self, other: Vec<ViewChangeProofSignature>, topology: &Topology) {
+        for (signatory_idx, signature) in other {
+            let public_key = topology.as_ref()[signatory_idx as usize].public_key();
+
+            if signature.verify(public_key, &self.payload).is_ok() {
+                self.signatures.push((signatory_idx, signature));
             }
         }
     }
```

Comment:
I don't know what the real harm will be, but attackers might be able to exchange each other's positions in the topology with the promise that they will exchange public keys and signatures and then declare them to the network:

```
topology: .., (i, key_j), .., (j, key_i), ..
other:    .., (i, sig_j), .., (j, sig_i), ..
```

this would pass the verification

Reply:
I posted just a thought to bring it to the attention of other reviewers. I'll try a little more thought.

I think this is an issue that appears only after unpacking signature. If there were any vulnerability, it would be in referencing the topology as a map from index to public key

Reply:
After thinking about it, such a situation is unlikely to occur since there would be no chance to misidentify one's public key to other peers after the peer registration